### DISCUSSION TITLE
path issues with matter-node.js and matter.js submodules

### DISCUSSION THREAD

**Discussion by vves** [matter-js/matter.js#75] [Q&A]
Working with the newly merged libraries, I am running into issues with the sub-module directory structure.  There doesn't seem to be a clear delineation of responsibilities in the submodules with cluster definitions in one place and the server for the cluster in the other sub-module.

Additionally, things like OnOffCluster exist in both submodules.

**What is the goal or intention of keeping matter-node.js and matter.js separate and what logic should exist in each if there is a good reason to keep them seperate?**

**Would it be easier to get the library back functional for those trying to leverage it and contribute to the clusters to merge the two sub-modules for the time being and then if the library gets unwieldy larger in the future, split it out then?**

It seems it would be easier to get contributors on the project if the structure were less complex and the library were useable via 'npm i <someGitLabRepo>' until the project is more mature and ready to be built for the public npm repo.

Maybe a Discord channel discussion is in order?

my 2 cents.
Wes

**Comment by vves** [comment]
Additionally - It seems that ALL the imports in the matter-node.js sub-module use the existing npm repo version of the `@project-chip/matter.js`.  Once both the sub-modules are deployed to the public npm repo. this should be a non-issue, but for anyone attempting to contribute by including the project-chip/matter.js library via GitHub npm installation will fail to resolve paths.   

!! Please advise if someone has a workaround for being able to run their own version of "Device.js" to test their clusters against live devices as the node-matter project allowed for this functionality.

**Reply by JimBuzbee** [reply]
@vves - I'm no expert and am not currently working on a cluster. But as a test I was able to build and run with a new dummy cluster like so (hopefully recalled correctly) :

> git clone https://github.com/project-chip/matter.js.git
> cd matter.js
> npm install
> edit packages/matter.js/src/cluster/ColorControlCluster.ts
>     (add skeleton implementation)
> edit packages/matter.js/src/matter.ts 
>       (add : export * from "./cluster/ColorControlCluster";)
> edit packages/matter-node.js/src/Device.ts
>     ( add references to my new cluster)
> npm run build
> node packages/matter-node.js/dist/Device.js
>

**Comment by Apollon77** [comment]
There should be nothing duplicated! So OnOffCluster definition is clearly onkly in matter.js, but yes the "OnOffHandler" is currently still in matter-node.js. This all will change when we move more code from matter-node.js to matter.js step by step in the next time. We have a fukl move concept at https://github.com/project-chip/matter.js/issues/62 which is in discussion.

That means, as of today, the Cluster definitions should all go into matter.js package. All rest regarding "implementation" matter-node.js is the better place now. We will make sure t move that over time into the correct place.
I will discuss with the other core devs if we might do the "big move" earlier then later, but in fact if yu are unsure where to put something new which is not a cluster definition, choose matter-node.js package! We also plan a npm release, but this do not help you if you also want to add functionality to matter.js.

It is correct that it is currently not that easy to develop "the library" AND also another package to use it. Here a temporary idea could be to create an own further package in the packages directory which would then, when using "npm i" on the matter.js root level should be bound correctly ... Yes you need to make sure to not commit that new package but for development time this should be most easy for now.
And yes it seems to be that all imports in matter-node.js use "npm @project-chip/mattr.js" but this is not true if ou do a "npm i" on package root level because then npm will locally create symlinks which end up in the fact that the local files are used. try it out!

I understand that this is not that easy for now and that we need to work on optimizing this ... and yes always lets chat in Discord if you have question!
